# Sowing Success: How Machine Learning Helps Farmers Select the Best Crops

![Farmer in a field](farmer_in_a_field.jpg)

Measuring essential soil metrics such as nitrogen, phosphorous, potassium levels, and pH value is an important aspect of assessing soil condition. However, it can be an expensive and time-consuming process, which can cause farmers to prioritize which metrics to measure based on their budget constraints.

Farmers have various options when it comes to deciding which crop to plant each season. Their primary objective is to maximize the yield of their crops, taking into account different factors. One crucial factor that affects crop growth is the condition of the soil in the field, which can be assessed by measuring basic elements such as nitrogen and potassium levels. Each crop has an ideal soil condition that ensures optimal growth and maximum yield.

A farmer reached out to you as a machine learning expert for assistance in selecting the best crop for his field. They've provided you with a dataset called `soil_measures.csv`, which contains:

- `"N"`: Nitrogen content ratio in the soil
- `"P"`: Phosphorous content ratio in the soil
- `"K"`: Potassium content ratio in the soil
- `"pH"` value of the soil
- `"crop"`: categorical values that contain various crops (target variable).

Each row in this dataset represents various measures of the soil in a particular field. Based on these measurements, the crop specified in the `"crop"` column is the optimal choice for that field.  

In this project, you will build multi-class classification models to predict the type of `"crop"` and identify the single most importance feature for predictive performance.

In [1]:
# All required libraries
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
#from sklearn.impute import SimpleImputer

# Load the dataset
crops = pd.read_csv("soil_measures.csv")

# Check for missing values in each feature column
crops.isna().sum()

N       0
P       0
K       0
ph      0
crop    0
dtype: int64

In [2]:
# Separate features and target variable
X = crops.drop("crop", axis=1)
y = crops["crop"]

In [3]:
# Split the dataset into training and testing sets
# X: features, y: target variable
# test_size=0.2 means 20% of the data will be used for testing, 80% for training
# random_state=12 ensures reproducibility of the split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=12, test_size=0.2)

In [4]:
features_dict = {}
feature_performance = {}
log_reg = LogisticRegression(multi_class="multinomial")

# Loop through each feature in the dataset
for feature in X.columns.tolist():
    print("Training with ", feature)
    
    # Train logistic regression using only the current feature
    log_reg.fit(X_train[[feature]], y_train)
    
    # Calculate accuracy score on the test set
    score = log_reg.score(X_test[[feature]], y_test)
    print("Score:", score)
    features_dict[feature] = score

    # Predict crop type using the current feature
    y_pred = log_reg.predict(X_test[[feature]])
    
    # Calculate weighted F1-score for the predictions
    feature_performance[feature] = metrics.f1_score(y_test, y_pred, average="weighted")
    print(f"F1-score for {feature}: {feature_performance[feature]}")

Training with  N
Score: 0.1318181818181818
F1-score for N: 0.09336902037284006
Training with  P
Score: 0.17727272727272728
F1-score for P: 0.10032552165844917
Training with  K
Score: 0.28863636363636364
F1-score for K: 0.2576237669707378
Training with  ph
Score: 0.07954545454545454
F1-score for ph: 0.043016625039667925


In [5]:
print("features_dict (accuracy for each feature):")
display(features_dict)

print("feature_performance (F1 score of each feature):")
display(feature_performance)

features_dict (accuracy for each feature):


{'N': 0.1318181818181818,
 'P': 0.17727272727272728,
 'K': 0.28863636363636364,
 'ph': 0.07954545454545454}

feature_performance (F1 score of each feature):


{'N': 0.09336902037284006,
 'P': 0.10032552165844917,
 'K': 0.2576237669707378,
 'ph': 0.043016625039667925}

In [6]:
# Find the feature with the highest F1-score
highest_feature_performance = max(feature_performance, key=feature_performance.get)

# Create a dictionary with the best feature and its F1-score
best_predictive_feature = {highest_feature_performance: feature_performance[highest_feature_performance]}

# Display the best predictive feature and its F1-score
best_predictive_feature

{'K': 0.2576237669707378}

---
## 🏁 Conclusions and Final Analysis

### 1. Key Findings
After evaluating each soil nutrient individually using **Logistic Regression**, the results showed that:
* **Potassium (K)** is the best individual predictor for crop type, achieving an **F1-score of 0.257**.
* **Phosphorous (P)** followed with a score of **0.100**.
* **Nitrogen (N)** and **pH** showed the lowest predictive power when used in isolation.

### 2. Technical Interpretation
The relatively low F1-scores across all individual features indicate that soil chemistry is complex. A single nutrient is not enough to accurately determine the ideal crop. This highlights the importance of **multivariate analysis** in agricultural data science.

### 3. Future Work & Improvements
To improve the model's accuracy, the following steps are proposed:
* **Multivariate Modeling:** Train a model using all features ($N, P, K, ph$) simultaneously.
* **Algorithm Benchmarking:** Compare Logistic Regression against non-linear models like **Random Forest** or **XGBoost**.
* **Data Scaling:** Apply `StandardScaler` to normalize the feature ranges.